src/biogeophys/SnowSnicarMod.F90

```diff
@@ -578,81 +792,236 @@ subroutine SNICAR_RT (flg_snw_ice, bounds, num_nourbanc, filter_nourbanc,  &
+                           AR_tmp = sno_AR(i)  ! user specified value
+                        endif
+                        do igb = 1, seven_bands
+                           g_ice_Cg_tmp(igb) = g_b0(igb) * ((fs_sphd/fs_hex)**g_b1(igb)) * (diam_ice**g_b2(igb))   ! Eq.7, He et al. (2017)
+                           gg_ice_F07_tmp(igb) = g_F07_c0(igb) + g_F07_c1(igb) * AR_tmp + g_F07_c2(igb) * (AR_tmp * AR_tmp)  ! Eqn. 3.1 in Fu (2007)
+                        enddo
+
+                     case ('hexagonal_plate')
+                        diam_ice = 2._r8 * snw_rds_lcl(i)   ! unit: microns
+                        if (sno_fs(i) == 0._r8) then
+                           fs_hex0 = fs_hex_ref  ! default; He et al. (2017), Table 1
+                        else
+                           fs_hex0 = sno_fs(i) ! user specified value
+                        endif
+                        fs_hex = fs_hex_ref  ! reference shape factor
+                        if (sno_AR(i) == 0._r8) then
+                           AR_tmp = AR_tmp_default_2  ! default; He et al. (2017), Table 1
+                        else
+                           AR_tmp = sno_AR(i)  ! user specified value
+                        endif
+                        do igb = 1, seven_bands
+                           g_ice_Cg_tmp(igb) = g_b0(igb) * ((fs_hex0/fs_hex)**g_b1(igb)) * (diam_ice**g_b2(igb))   ! Eq.7, He et al. (2017)
+                           gg_ice_F07_tmp(igb) = g_F07_p0(igb) + g_F07_p1(igb) * log(AR_tmp) + g_F07_p2(igb) * (log(AR_tmp) * log(AR_tmp)) ! Eqn. 3.3 in Fu (2007)
+                        enddo
+
+                     case ('koch_snowflake')
+                        diam_ice = 2._r8 * snw_rds_lcl(i) / 0.544_r8  ! unit: microns
+                        if (sno_fs(i) == 0._r8) then
+                           fs_koch = fs_koch_default  ! default; He et al. (2017), Table 1
+                        else
+                           fs_koch = sno_fs(i) ! user specified value
+                        endif
+                        fs_hex = fs_hex_ref  ! reference shape factor
+                        if (sno_AR(i) == 0._r8) then
+                           AR_tmp = AR_tmp_default_2  ! default; He et al. (2017), Table 1
+                        else
+                           AR_tmp = sno_AR(i)  ! user specified value
+                        endif
+                        do igb = 1, seven_bands
+                           g_ice_Cg_tmp(igb) = g_b0(igb) * ((fs_koch/fs_hex)**g_b1(igb)) * (diam_ice**g_b2(igb))   ! Eq.7, He et al. (2017)
+                           gg_ice_F07_tmp(igb) = g_F07_p0(igb) + g_F07_p1(igb) * log(AR_tmp) + g_F07_p2(igb) * (log(AR_tmp) * log(AR_tmp)) ! Eqn. 3.3 in Fu (2007)
+                        enddo
+
+                     case default
+                        write(iulog,*) subname//' ERROR: unknown sno_shp for i: ', sno_shp(i), i
+                        call endrun(msg=errMsg(sourcefile, __LINE__))
+                     end select
```

Comment:
@cenlinhe I find that the model stops with an error when I run with this combination:
```
SNICAR: downward solar radiation spectrum type =mid_latitude_winter
SNICAR: dust optics type = sahara
SNICAR: number of bands in snow albedo calculation =           5
SNICAR: snow grain shape type = sphere
SNICAR: BC-snow internal mixing =  F
SNICAR: dust-snow internal mixing =  F
SNICAR: OC in snow =  F
Number of snow layers =      12
```
(Until now I always ran with hexagonal_plate, which worked.)

The error says (line 886, see above):
` SNICAR_RT ERROR: unknown sno_shp for i: sphere                   -9`
...where i is the snow layer. Do you have a suggestion, or should we disable "sphere"?

Reply:
Let me clarify that I added this error check in a recent round of revisions.
@cenlinhe I see your comment in line 826, "Nonspherical snow: ..."
So maybe adding this error check was a mistake.
Let me know what you think and meanwhile I will remove and retest.

Reply:
I did not know see this model crash using my original code. I know what's going on here. These calculations within the "case select" only apply for nonspherical shape, so when the "sphere" is selected, this part is automatically skipped. So please do not put a model stop for "case default". You could put some print-out message saying the sphere is selected, but not let the model stop.